Add clippy to CI

marcelstanley opened the pull request:
#### TL;DR:

**The changes made here are up for discussion**, including the commit messages themselves.

Most of the warnings reported by Clippy were fixed. Besides that, there are some warnings that were suppressed (see `chore` commits) via configuration, mostly because fixing them might take more time.

---

The `ubuntu-latest` image used by CI relies on an outdated version of `clippy`, which makes [the linter fail](https://github.com/cartesi/rollups/actions/runs/5556711553/jobs/10149567440) due to a convoluted [*clone_double_ref* error](https://users.rust-lang.org/t/clone-double-ref/92612), which has been dropped in the most recent `clippy` version [1].

By simply updating `rustup` [1], [clippy passes](https://github.com/cartesi/rollups/actions/runs/5556801966/jobs/10149774799), even though there are several warnings, which are mostly addressed by the commits in this PR.

In [the latest CI run](https://github.com/cartesi/rollups/actions/runs/5557061014/jobs/10150367811?pr=159), with all changes made, there are still the following outstanding warnings:

- [*new_ret_no_self*](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) - either not applicable or costly to fix due to the current business logic;
- [*module_inception*](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) - affects code generated via `state-fold-types` and could not be suppressed because the suggested configuration variable did not work under such condition.

---

[1] Expected `rust`-related versions:
```shell
$ rustup -V && rustc -V && cargo -V && cargo clippy -V
rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.71.0 (8ede3aae2 2023-07-12)`
rustc 1.71.0 (8ede3aae2 2023-07-12)
cargo 1.71.0 (cfd3bbd8f 2023-06-08)
clippy 0.1.71 (8ede3aa 2023-07-12)
``` 

---

Closes cartesi/rollups#25

marcelstanley committed:
ci(offchain): run code quality assessment

While here, move code format and dependency checking to the code quality
workflow.

marcelstanley committed:
fix(offchain): apply fixes suggested by rust-clippy

Full list of warnings:

- https://rust-lang.github.io/rust-clippy/master/index.html#redundant_pattern_matching
- https://rust-lang.github.io/rust-clippy/master/index.html#/needless_borrow
- https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_static_lifetimes
- https://rust-lang.github.io/rust-clippy/master/index.html#/map_flatten
- https://rust-lang.github.io/rust-clippy/master/index.html#/suspicious_doc_comments
- https://rust-lang.github.io/rust-clippy/master/index.html#/assign_op_pattern
- https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
- https://rust-lang.github.io/rust-clippy/master/index.html#/clone_on_copy
- https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern
- https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
- https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
- https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
- https://rust-lang.github.io/rust-clippy/master/index.html#to_string_in_format_args
- https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark
- https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
- https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
- https://rust-lang.github.io/rust-clippy/master/index.html#redundant_field_names
- https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
- https://rust-lang.github.io/rust-clippy/master/index.html#/comparison_to_empty
- https://rust-lang.github.io/rust-clippy/master/index.html#/unnecessary_cast

marcelstanley committed:
chore(offchain): suppress rust-clippy warnings

Full list of warnings:
- avoid warning `large_enum_variant` from cargo clippy on `pub enum MachineOneof`, built from grpc-interfaces. More at https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
- allow enum variants with same suffix. More at https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names.
- allow too many arguments in fn. More at https://rust-lang.github.io/rust-clippy/master/index.html#/too_many_arguments.
- allow capitalized acronym. More at https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms.

torives reviewed (commented):
Regarding the remaining warnings, I propose we create an issue for each and postpone promoting warnings to errors on the CI until we fix them.

The `state-fold` code will soon share the same repo as the other services, so the [module_inception](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) issue can be tackled then. And we can discuss how to fix [new_ret_no_self](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) in its own issue.

We'll squash the commits fixing the lint errors/warnings into one, right?

marcelstanley commented on .github/workflows/code-quality.yml:
> Regarding the remaining warnings, I propose we create an issue for each and postpone promoting warnings to errors on the CI until we fix them.
> 
Nice, that's the idea!

> The `state-fold` code will soon share the same repo as the other services, so the [module_inception](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) issue can be tackled then. And we can discuss how to fix [new_ret_no_self](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) in its own issue.
> 
Great!

> We'll squash the commits fixing the lint errors/warnings into one, right?
Yes, we may. And this will help with keeping consistency for the conventional commits initiative, even though I've preferred splitting them to guide the review process.

> What do you think of also moving the `rustfmt` step to this workflow?
I agree!

marcelstanley replied:
Moved code format task at [92de1c7](https://github.com/cartesi/rollups/pull/159/commits/92de1c78f6e4caea99ac7bd9e49143d7b316fe9d)

torives replied:
The split commits made the review a lot easier btw, thx <3

marcelstanley commented:
Rebases changes on top of `next`

marcelstanley committed:
ci(offchain): disable lints for deferred warnings

The following outstanding warnings were suppressed while we amend the code
to meet the linter requirements:

- https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self
- https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception

marcelstanley commented:
@cartesi/node-unit, lints [*new_ret_no_self*](https://rust-lang.github.io/rust-clippy/master/index.html#/new_ret_no_self) and [*module_inception*](https://rust-lang.github.io/rust-clippy/master/index.html#/module_inception) were disabled at 68447d7571eb086939c4cc065cc7aa0d40893a80, as neither of them could be supressed via configuration. We should work on fixing the related warnings later on at #202.

Besides that, I'd like to highlight once again that a few other lints have been supressed (see commits below) for various reasons:

- 98370b1929191827a25b37f7d9cbfd52ac646161
- 4d518da7793619fda86fadcf626921419b53ba6b
- 2fdd9161461cfecf8b50ef4a9da91fe0d88e4b8e
- 993d7e2edd5443cc29ddc2be4c1f5b648e7818df

Notice, though, that all of them could have been addressed with somewaht more complex changes in the code and could be part of the forthcoming task #202 as well.

torives commented:
https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb: I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from.
https://github.com/cartesi/rollups/commit/d00f2f13676e6f482e84cce4b0a35acaa9a52e67: Maybe I'm just used to it, but I have no issues with "Error" as a suffix of error types.
https://github.com/cartesi/rollups/commit/93ac098f3760d636228c70bfcf6d6c4b8aec6486: This is a good candidate for a refactor and IMO should be included in #202
https://github.com/cartesi/rollups/commit/02f79f93eb9c4907793bd49a5ec8aa6849d3097e: No strong opinions on this one, but I'd follow the linter, just because.

gligneul reviewed (approved):
Looks good. Do you plan to squash the commits?

marcelstanley commented:
> Looks good. Do you plan to squash the commits?

Yes, I will squash all fixes into one.

marcelstanley commented:
> [624cb11](https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb): I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from. 

Done at 5226b24e

marcelstanley commented:
> [624cb11](https://github.com/cartesi/rollups/commit/624cb110e5ab5f6b3a96b852b496f4d7ac7f00eb): I don't think there's much we can do as this is auto-generated from an external dependency. I would leave it as it is, maybe adding a comment to the specific line in the `clippy.toml` file, so we know where that came from. 

Done at 5226b24e, @torives.
